service/access.go

```diff
@@ -0,0 +1,395 @@
+	glg.Info("refreshAccessTokenCache started")
+
+	echan := make(chan error, a.tokenCache.Len()*(a.errRetryMaxCount+1))
+	go func() {
+		defer close(echan)
+
+		a.tokenCache.Foreach(ctx, func(key string, val interface{}, exp int64) bool {
+			domain, role, principal := decode(key)
+			cd := val.(*accessCacheData)
+
+			for err := range a.updateAccessTokenWithRetry(ctx, domain, role, principal, cd.expiresIn) {
+				echan <- err
+			}
+			return true
+		})
+	}()
+
+	return echan
+}
+
+func (a *accessService) updateAccessTokenWithRetry(ctx context.Context, domain, role, proxyForPrincipal string, expiresIn int64) <-chan error {
+	glg.Debugf("updateAccessTokenWithRetry started, domain: %s, role: %s, proxyForPrincipal: %s, expiresIn: %d", domain, role, proxyForPrincipal, expiresIn)
+
+	echan := make(chan error, a.errRetryMaxCount+1)
+	go func() {
+		defer close(echan)
+
+		for i := 0; i <= a.errRetryMaxCount; i++ {
+			if _, err := a.updateAccessToken(ctx, domain, role, proxyForPrincipal, expiresIn); err != nil {
+				echan <- err
+				time.Sleep(a.errRetryInterval)
+			} else {
+				glg.Debug("update success")
+				break
+			}
+		}
+	}()
+
+	return echan
+}
+
+// updateAccessToken returns AccessTokenResponse struct or error.
+// This function ask athenz to generate access token and return, or return any error when generating the access token.
+func (a *accessService) updateAccessToken(ctx context.Context, domain, role, proxyForPrincipal string, expiresIn int64) (*AccessTokenResponse, error) {
+	key := encode(domain, role, proxyForPrincipal)
+	expTimeDelta := fastime.Now().Add(time.Minute)
+
+	at, err, _ := a.group.Do(key, func() (interface{}, error) {
+		at, e := a.fetchAccessToken(ctx, domain, role, proxyForPrincipal, expiresIn)
+		if e != nil {
+			return nil, e
+		}
+
+		a.tokenCache.SetWithExpire(key, &accessCacheData{
+			token:             at,
+			domain:            domain,
+			role:              role,
+			proxyForPrincipal: proxyForPrincipal,
+			expiresIn:         expiresIn,
+		}, time.Unix(at.ExpiresIn, 0).Sub(expTimeDelta))
+
+		glg.Debugf("token is cached, domain: %s, role: %s, proxyForPrincipal: %s, expiry time: %v", domain, role, proxyForPrincipal, at.ExpiresIn)
+		return at, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return at.(*AccessTokenResponse), err
+}
+
+// fetchAccessToken fetches the access token from Athenz server, and returns the AccessTokenResponse or any error occurred.
+func (a *accessService) fetchAccessToken(ctx context.Context, domain, role, proxyForPrincipal string, expiry int64) (*AccessTokenResponse, error) {
+	glg.Debugf("get access token, domain: %s, role: %s, proxyForPrincipal: %s, expiry: %d", domain, role, proxyForPrincipal, expiry)
+
+	// get the n-token
+	cred, err := a.token()
+	if err != nil {
+		return nil, err
+	}
+
+	// create scope
+	glg.Debugf("roleSeparater: %s, scopeSeparater: %s", roleSeparater, scopeSeparater)
+	scope := createScope(domain, role)
```

Comment:
maybe we need a debug message for the `scope` variable.

Reply:
I agree, nice suggest.

c01882f